src/de.rs

```diff
@@ -2155,16 +2155,18 @@ where
         }
     }
 
-    deserialize_integer_key!(deserialize_i8 => visit_i8);
-    deserialize_integer_key!(deserialize_i16 => visit_i16);
-    deserialize_integer_key!(deserialize_i32 => visit_i32);
-    deserialize_integer_key!(deserialize_i64 => visit_i64);
-    deserialize_integer_key!(deserialize_i128 => visit_i128);
-    deserialize_integer_key!(deserialize_u8 => visit_u8);
-    deserialize_integer_key!(deserialize_u16 => visit_u16);
-    deserialize_integer_key!(deserialize_u32 => visit_u32);
-    deserialize_integer_key!(deserialize_u64 => visit_u64);
-    deserialize_integer_key!(deserialize_u128 => visit_u128);
+    deserialize_numeric_key!(deserialize_i8 => visit_i8);
+    deserialize_numeric_key!(deserialize_i16 => visit_i16);
+    deserialize_numeric_key!(deserialize_i32 => visit_i32);
+    deserialize_numeric_key!(deserialize_i64 => visit_i64);
+    deserialize_numeric_key!(deserialize_i128 => visit_i128);
+    deserialize_numeric_key!(deserialize_u8 => visit_u8);
+    deserialize_numeric_key!(deserialize_u16 => visit_u16);
+    deserialize_numeric_key!(deserialize_u32 => visit_u32);
+    deserialize_numeric_key!(deserialize_u64 => visit_u64);
+    deserialize_numeric_key!(deserialize_u128 => visit_u128);
+    deserialize_numeric_key!(deserialize_f32 => visit_f32);
+    deserialize_numeric_key!(deserialize_f64 => visit_f64);
```

Comment:
`.parse()` is not appropriate to use here for floats. That accepts a bunch of syntax that is not considered a valid float in json. For example `.0` or `00`. Please make sure this accepts only representations that would be valid f32/f64 outside of a map key.

Reply:
I've updated this to delegate to the parent `Deserializer` via `self.de.method(...)`. This unfortunately makes the error messages a bit worse, as we now get something like:

```
expected value at line 1 column 3
```

Instead of:

```
"invalid type: string \"x\", expected f32 at line 1 column 4"
```

I noticed a similar issue here is present with integer keys, [this example](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=9d8c47583ecfb7e0baf8306fdd082f50) works on `serde_json` `master`:

```
extern crate serde_json;
extern crate serde;

fn main() {
    dbg!(serde_json::from_str::<std::collections::HashMap<i32, i32>>(r#"{"00":0}"#));
}
```

Is it worth updating this as well?

Reply:
Yes, good call. I filed #1034 to follow up.

Reply:
In an effort to patch #1034, I've also updated integer key parsing to use `Deserialize::*`. This still exhibits the above error degredation.

Would it make sense here to parse the key into an intermediate string buffer, and then deserialize the integer value using `crate::from_str`? This would result in the original error messages for unexpected string keys, but may have a performance impact.

Reply:
This needs a check for leading whitespace, otherwise it would end up accepting potentially invalid JSON syntax.

For example this is not a legal JSON document:

```json
{"
1":null}
```

```rust
// [dependencies]
// ordered-float = { version = "3", features = ["serde"] }
// serde_json = "1"

use ordered_float::NotNan;
use std::collections::BTreeMap;

fn main() {
    let invalid_json = r#"
        {"
        1":null}
    "#;

    let map: BTreeMap<NotNan<f64>, ()> = serde_json::from_str(invalid_json).unwrap();
    println!("{:?}", map);
}
```

Reply:
That'll fix the error message problem too. JSON numbers must start with `0-9` or `-`. If the upcoming character is not one of those, you can avoid calling the function that would produced the bad message in this situation, and directly create a useful "invalid value" error instead.

Reply:
I've patched this with a check for whitespace characters. I don't quite follow what you mean about fixing the error, the misleading text is still present, no? E.g. this example:

```rust
fn main() {
    dbg!(serde_json::from_str::<std::collections::HashMap<i32, i32>>(r#"{"abc":0}"#));
}
```

Previously would print `invalid type: string "abc", expected i32 at line 1 column 6`, but after this changeset it prints `expected value at line 1 column 3`.

Reply:
This is what I was thinking of doing to give a more meaningful error: #1035.